File: app/src/main/java/com/adityakamble49/wordlist/ui/common/WordListViewModelFactory.java
package com.adityakamble49.wordlist.ui.common;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import java.util.Map;

import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;

/**
 * Word List View Model Factory
 *
 * @author Aditya Kamble
 * @since 30/9/2018
 */
@Singleton
public class WordListViewModelFactory implements ViewModelProvider.Factory {
    private final Map<Class<? extends ViewModel>, Provider<ViewModel>> creators;

    @Inject
    public WordListViewModelFactory(Map<Class<? extends ViewModel>, Provider<ViewModel>> creators) {
        this.creators = creators;
    }

    @SuppressWarnings("unchecked")
    @Override
    public <T extends ViewModel> T create(Class<T> modelClass) {
        Provider<? extends ViewModel> creator = creators.get(modelClass);
        if (creator == null) {
            for (Map.Entry<Class<? extends ViewModel>, Provider<ViewModel>> entry : creators.entrySet()) {
                if (modelClass.isAssignableFrom(entry.getKey())) {
                    creator = entry.getValue();
                    break;
                }
            }
        }
        if (creator == null) {
            throw new IllegalArgumentException("unknown model class " + modelClass);
        }
        try {
            return (T) creator.get();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: app/src/main/java/com/adityakamble49/wordlist/ui/list/WordListViewModelFactory.kt
package com.adityakamble49.wordlist.ui.list

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import com.adityakamble49.wordlist.cache.PreferenceHelper
import com.adityakamble49.wordlist.cache.db.WordRepo

/**
 * @author Aditya Kamble
 * @since 4/4/2018
 */
class WordListViewModelFactory(
        private val preferenceHelper: PreferenceHelper,
        private val wordRepo: WordRepo) : ViewModelProvider.Factory {

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(WordListViewModel::class.java)) {
            return WordListViewModel(preferenceHelper, wordRepo) as T
        }
        throw IllegalArgumentException("Unknown ViewModel Class")
    }
}